### PR TITLE
Improve credential chain error reporting and skip non-applicable providers

### PR DESCRIPTION
**Pull request by ysaito1001**
## Motivation and Context
When the credential chain fails, the error surfaced to the caller is a generic `DispatchFailure(... an error occurred while loading credentials ...)`; discovering which provider failed and why requires re-running with `RUST_LOG=aws_config=debug`. In addition, when `AWS_PROFILE` refers to a profile missing from `~/.aws/config`, the Rust SDK hard-fails the chain, while other SDKs (e.g., Java, .NET) skip and continue. An upcoming design recommends that providers whose source is not detected do not participate in the default credentials chain, whereas providers whose source was detected and mark themselves "terminal" participate in the chain but terminate fall-through at runtime if identity resolution fails.

## Description
The upcoming design proposes to make the default credentials chain modular by separating chain construction from identity resolution. At construction time, providers register resolvers only when their source is detected via cheap signals (e.g., env var is set, profile declares the source). At resolution time, the chain iterates the assembled resolver list. The current Rust SDK collapses these two phases into one, using the error type (`CredentialsNotLoaded` vs other) at runtime as a proxy for whether a provider should have been in the chain at all.

This PR is an interim fix that aligns the Rust SDK's behavior with the future design's semantics without implementing the full construction/resolution separation yet.

- Returning `CredentialsNotLoaded` for non-applicable providers
Providers whose source isn't detected now return `CredentialsNotLoaded` instead of `InvalidConfiguration`, allowing the chain to continue. This affects three cases: the profile provider when the selected profile doesn't exist (`MissingProfile`) or is configured for the token chain rather than the credentials chain (`TokenProviderConfig`), and the web identity token provider when `AWS_WEB_IDENTITY_TOKEN_FILE` is set but `AWS_ROLE_ARN` is missing (incomplete detection signal). In the future design, these providers would simply not register during construction.
  
- Adding `ProviderChainError<E>` with per-provider summaries
When all providers are exhausted, the error now includes a structured summary showing what each provider reported. For instance, the following output from a unit test `error_message_includes_per_provider_summary` shows a chain where all providers were skipped:
```
  no credentials found in chain. Attempted:
    Environment: the credential provider was not enabled: not set
    Profile: the credential provider was not enabled: profile 'deploy' not found
    IMDS: the credential provider was not enabled: could not communicate with IMDS
```

Note hard-fail semantics are preserved: non-`CredentialsNotLoaded` errors (e.g., STS returned `InvalidIdentityTokenException`, expired SSO token) still stop the chain immediately. This corresponds to the future design's terminal resolver behavior; the source was detected, the provider is authoritative, and failure is not masked by falling through.

**Question**: Do we need a `BehaviorVersion` gate for the change? My take is no; the change makes the chain more permissive (fall through instead of hard-fail), which is a bug fix aligning with other SDKs rather than a behavior customers intentionally depend on. Open to what reviewers think.

## Testing
- CI
- Four new unit tests cover the exhausted error message format, hard-fail behavior, empty chain edge case, and programmatic downcast access to `ProviderChainError` (A convenience accessor on `CredentialsError` would create a circular dependency because the error lives in `aws-credential-types` while `ProviderChainError` lives in `aws-config`, where the chain concept belongs.)
- The existing `web_identity_token::test::missing_env_var` assertion was updated to expect `CredentialsNotLoaded` instead of `InvalidConfiguration`.

## Checklist
- [x] For changes to the AWS SDK, generated SDK code, or SDK runtime crates, I have created a changelog entry Markdown file in the `.changelog` directory, specifying "aws-sdk-rust" in the `applies_to` key.

----

_By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice._

### CODE REVIEW

File: aws/rust-runtime/aws-config/src/meta/mod.rs [hunk 1]
```diff
@@ -5,6 +5,87 @@
...
+
+impl<E: Error> fmt::Display for ProviderChainError<E> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        if self.attempts.is_empty() {
+            return write!(f, "no providers were configured in the chain");
+        }
+        write!(f, "no credentials found in chain. Attempted:")?;
+        for attempt in &self.attempts {
+            write!(f, "\n  {}: {}", attempt.name, attempt.error)?;
+            let mut source = attempt.error.source();
```
Comment (landonxjames) [inline aws/rust-runtime/aws-config/src/meta/mod.rs:81]: I assume errors should never contain the secrets used to attempt to get the credentials so we are ok not obfuscating this?

**Review by landonxjames — APPROVED** [review]
LGTM!

File: aws/rust-runtime/aws-config/src/meta/mod.rs (hunk 1 above)
Comment (ysaito1001) [reply aws/rust-runtime/aws-config/src/meta/mod.rs:81]: Correct, the error chain only contains diagnostic context (e.g., provider names, config locations, and connectivity messages), but not credential values. The existing providers ensure that secrets don't make it into `Error::source()`.
